src/react/elements.js

```diff
@@ -15,38 +15,158 @@ import {
   AbstractValue,
   AbstractObjectValue,
   ArrayValue,
+  FunctionValue,
   NullValue,
   NumberValue,
   ObjectValue,
+  StringValue,
   Value,
 } from "../values/index.js";
 import { Create } from "../singletons.js";
 import invariant from "../invariant.js";
 import { Get } from "../methods/index.js";
 import {
   applyObjectAssignConfigsForReactElement,
+  cloneObject,
   createInternalReactElement,
   flagPropsWithNoPartialKeyOrRef,
-  hardModifyReactObjectPropertyBinding,
   getProperty,
+  hardModifyReactObjectPropertyBinding,
   hasNoPartialKeyOrRef,
+  isEventProp,
+  transferSafePropertiesToRemoveFromObjectsToProps,
 } from "./utils.js";
 import * as t from "babel-types";
 import { computeBinary } from "../evaluators/BinaryExpression.js";
 import { CompilerDiagnostic, FatalError } from "../errors.js";
 
+function markToSafelyRemovePropertyDuringSerialization(realm: Realm, obj: ObjectValue, propName: string): void {
+  let propsToRemove = realm.react.objectsWithPropsToRemove.get(obj);
+
+  if (propsToRemove === undefined) {
+    propsToRemove = new Set();
+    realm.react.objectsWithPropsToRemove.set(obj, propsToRemove);
+  }
+  propsToRemove.add(propName);
+}
+
+function deepTraverseAndFindOrDeleteFirstRenderProperties(
+  realm: Realm,
+  obj: ObjectValue | AbstractObjectValue,
```

Comment:
Can we call this `propsObj` or something to indicate this doesn't actually deeply traverse arbitrary objects? I didn't initially catch that.